File: exercism/python/list-ops/list_ops.py
"""
List Ops
"""


def append(list1, list2):
    """
    `append` (*given two lists, add all items in the second list to the end of the first list*);
    """
    return list1 + list2


def concat(lists):
    """
    `concatenate` (*given a series of lists, combine all items in all lists
        into one flattened list*);
    """
    return [item for sublist in lists for item in sublist]


def filter(function, list):
    """
    `filter` (*given a predicate and a list, return the list of all items for
        which `predicate(item)` is True*);
    """
    return [item for item in list if function(item)]


def length(list):
    """
    `length` (*given a list, return the total number of items within it*);
    """
    return len(list)


def map(function, list):
    """
    `map` (*given a function and a list, return the list of the results of
        applying `function(item)` on all items*);
    """
    return [function(item) for item in list]


def foldl(function, list, initial):
    """
    `foldl` (*given a function, a list, and initial accumulator, fold (reduce)
        each item into the accumulator from the left using `function(accumulator, item)`*);
    """
    for item in list:
        initial = function(initial, item)
    return initial


def foldr(function, list, initial):
    """
    `foldr` (*given a function, a list, and an initial accumulator, fold (reduce)
        each item into the accumulator from the right using `function(item, accumulator)`*);
    """
    for item in list[::-1]:
        initial = function(item, initial)
    return initial


def reverse(list):
    """
    `reverse` (*given a list, return a list with all the original items, but in reversed order*);
    """
    return list[::-1]

File: exercism/go/list-ops/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

// type binFunc func(x, y int) int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	for _, iten := range s {
		initial = fn(initial, iten)
	}
	return initial
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	for i := len(s) - 1; i >= 0; i-- {
		initial = fn(s[i], initial)
	}
	return initial
}

func (s IntList) Filter(fn func(int) bool) IntList {
	list := IntList{}
	for _, iten := range s {
		if fn(iten) {
			list = append(list, iten)
		}
	}
	return list
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	list := IntList{}
	for _, iten := range s {
		list = append(list, fn(iten))
	}
	return list
}

func (s IntList) Reverse() IntList {
	list := IntList{}
	for i := len(s) - 1; i >= 0; i-- {
		list = append(list, s[i])
	}
	return list
}

func (s IntList) Append(lst IntList) IntList {
	// Append one slice to another
	return append(s, lst...)
	// The ... unpacks b. Without the dots, the code would
	// attempt to append the slice as a whole, which is invalid.
}

func (s IntList) Concat(lists []IntList) IntList {
	for _, iten := range lists {
		s = append(s, iten...)
	}
	return s
}
